NebulaWorld/Logistics/StationUIManager.cs

```diff
@@ -0,0 +1,464 @@
+
+        public static void Initialize()
+        {
+            StationUISubscribers = new Dictionary<int, List<NebulaConnection>>();
+            UpdateCooldown = 0;
+            UIIsSyncedStage = 0;
+            UIStationId = 0;
+            UIRequestedShipDronWarpChange = false;
+        }
+        public static void AddSubscriber(int stationGId, NebulaConnection player)
+        {
+            List<NebulaConnection> players;
+            if(!StationUISubscribers.TryGetValue(stationGId, out players))
+            {
+                players = new List<NebulaConnection>();
+                StationUISubscribers.Add(stationGId, players);
+            }
+            for (int i = 0; i < StationUISubscribers[stationGId].Count; i++)
+            {
+                if(StationUISubscribers[stationGId][i] == player)
+                {
+                    return;
+                }
+            }
+            StationUISubscribers[stationGId].Add(player);
+        }
+        public static void RemoveSubscriber(int stationGId, NebulaConnection player)
+        {
+            List<NebulaConnection> players;
+            if(StationUISubscribers.TryGetValue(stationGId, out players))
+            {
+                StationUISubscribers[stationGId].Remove(player);
+                if(StationUISubscribers[stationGId].Count == 0)
+                {
+                    StationUISubscribers.Remove(stationGId);
+                }
+            }
+        }
+        public static List<NebulaConnection> GetSubscribers(int stationGId)
+        {
+            List<NebulaConnection> players;
+            if(!StationUISubscribers.TryGetValue(stationGId, out players))
+            {
+                return new List<NebulaConnection>();
+            }
+            return players;
+        }
+
+        public static void DecreaseCooldown()
+        {
+            // cooldown is for the storage sliders
+            if(UpdateCooldown > 0)
+            {
+                UpdateCooldown--;
+            }
+        }
+
+        public static void UpdateUI(StationUI packet)
+        {
+            if((UpdateCooldown == 0 || !packet.isStorageUI) && LocalPlayer.IsMasterClient)
+            {
+                UpdateCooldown = 10;
+                if (packet.isStorageUI)
+                {
+                    UpdateStorageUI(packet);
+                }
+                else
+                {
+                    UpdateSettingsUI(packet);
+                }
+            }
+            else if(!LocalPlayer.IsMasterClient)
+            {
+                if (packet.isStorageUI)
+                {
+                    UpdateStorageUI(packet);
+                }
+                else
+                {
+                    UpdateSettingsUI(packet);
+                }
+            }
+        }
+
+        private static void UpdateSettingsUIBackground(StationUI packet, PlanetData pData, int stationGId)
```

Comment:
This method seems heavy and it looks like there is some code copied from the game. We should probably look if we can simplify it or call the actual game code instead.

Reply:
I am calling the vanilla game functions in `UpdateSettingsUI()`, but since they belong to `UIStationWindow` i cant call them in `UpdateSettingsUIBackground()` since the window is closed at that time.

I could use reverse patching to get a modified version of the original method which i then could use there, but in the end it would be more complicated code including transpilers to get the same done.

I can see how i can make the method a bit easier to read.
basically its the functionality of 
![grafik](https://user-images.githubusercontent.com/79335616/116382437-32cd6300-a816-11eb-995e-766ab845dc12.png)